command/node/drain.go

```diff
@@ -54,14 +69,127 @@ func Drain(c *cli.Context, logger *log.Logger) error {
+			nodeAllocations, _, err := nomadClient.Nodes().Allocations(node.ID, nil)
+			if err != nil {
+				log.Errorf("Error updating scheduling eligibility for %s: %s", node.Name, err)
+				continue
+			}
+
+			for _, nodeAllocation := range nodeAllocations {
+				if *nodeAllocation.Job.Type != nomadStructs.JobTypeService {
+					log.Infof("Skipping %s because it's not a service job", nodeAllocation.JobID)
+					continue
+				}
+
+				if nodeAllocation.ClientStatus == nomadStructs.AllocClientStatusComplete || nodeAllocation.DesiredStatus == nomadStructs.AllocDesiredStatusStop {
+					log.Infof("Skipping %s because it's already complete", nodeAllocation.JobID)
+					continue
+				}
+
+				log.Infof("Found Allocation %s, for job %s, moving it", nodeAllocation.ID, nodeAllocation.JobID)
+
+				allocationJob := nodeAllocation.Job
+				existingConstraintAppended := false
+				for taskGroupIndex, taskGroup := range allocationJob.TaskGroups {
+					if *taskGroup.Name == nodeAllocation.TaskGroup {
+						for constraintIndex, constraint := range taskGroup.Constraints {
+							if constraint.LTarget == newConstraint.LTarget {
+								allocationJob.TaskGroups[taskGroupIndex].Constraints[constraintIndex] = newConstraint
+								existingConstraintAppended = true
+							}
+						}
+						if !existingConstraintAppended {
+							allocationJob.TaskGroups[taskGroupIndex].Constrain(newConstraint)
+						}
+					}
+				}
+				registerResponse, _, err := nomadClient.Jobs().Register(allocationJob, nil)
+				if err != nil {
+					return fmt.Errorf("failed to move taskgroup %s for job %s: %s", nodeAllocation.TaskGroup, nodeAllocation.JobID, err)
+				}
+				if c.Bool("wait-for-pending") {
+					log.Infof("Waiting for successfully placing the moved job")
+					// wait for the evaluation to be available
+					for {
+						time.Sleep(1 * time.Second)
+						evaluation, _, err := nomadClient.Evaluations().Info(registerResponse.EvalID, nil)
+						if err != nil {
+							continue
+						}
+
+						if evaluation.Status == nomadStructs.EvalStatusCancelled || evaluation.Status == nomadStructs.EvalStatusFailed {
+							logger.Errorf("Could not evaluate the job: %s", evaluation.StatusDescription)
+							continue
+						}
+
+						if evaluation.Status == nomadStructs.EvalStatusComplete {
+							log.Infof("Evaluation %s for job %s completed", evaluation.ID, nodeAllocation.JobID)
+							break
+						}
+					}
+
+					// wait for the allocations to be available
+					for {
+						time.Sleep(1 * time.Second)
+						evaluations, _, err := nomadClient.Evaluations().List(nil)
+						if err != nil {
+							continue
+						}
+						for _, evaluation := range evaluations {
+							if &evaluation.JobID == allocationJob.ID && evaluation.Status == nomadStructs.EvalStatusBlocked {
+								log.Infof("Job %s got blocked evaluations", allocationJob.ID)
+								continue
+							}
+						}
+
+						break
+					}
+
+					// waiting for allocation to be placed
```

Comment:
much rather see a function that filters the queried allocation list by job id, and then once you have the list of all possible allocations from it, then you have a goroutine that runs against each of the found allocs to query for task state.

Reply:
If we'll call allocation list only once and placing allocation is failed (docker pull timeout) we will miss the new allocation evaluated for this job

Reply:
makes sense